File: game/src/test/java/com/bship/games/endpoints/cabinet/entity/BoardTest.java
package com.bship.games.endpoints.cabinet.entity;

import com.bship.games.logic.definitions.Direction;
import com.bship.games.logic.definitions.Harbor;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExpectedException;

import static java.util.Collections.singletonList;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;

public class BoardTest {

    @Rule
    public ExpectedException thrown = ExpectedException.none();
    private Piece battleship;
    private Piece aircraftCarrier;

    @Before
    public void setup() {
        Point start = new Point(0, 0);
        Harbor type = Harbor.BATTLESHIP;
        battleship = Piece.builder()
                .withType(type)
                .withPlacement(start)
                .withOrientation(Direction.DOWN)
                .build();

        Point start1 = new Point(1, 0);
        Harbor type1 = Harbor.AIRCRAFT_CARRIER;
        aircraftCarrier = Piece.builder()
                .withType(type1)
                .withPlacement(start1)
                .withOrientation(Direction.DOWN)
                .build();
    }

    @Test
    public void addShip_shouldBeAbleToAddAShipToTheBoard() {
        Board board = Board.builder().addPiece(battleship).build();

        assertThat(board.getPieces(), contains(battleship));
    }

    @Test
    public void addShip_shouldBeAbleToAddMultipleShipsToTheBoard() {
        Board board = Board.builder().addPiece(battleship).build();
        Board newBoard = board.copy().addPiece(aircraftCarrier).build();

        assertThat(newBoard.getPieces(), contains(battleship, aircraftCarrier));
    }

    @Test
    public void getShips__with_addShip_shouldBeImmutable() {
        thrown.expect(UnsupportedOperationException.class);

        Board board = Board.builder().addPiece(battleship).build();

        board.getPieces().add(aircraftCarrier);
    }

    @Test
    public void getShips__with_withShips_shouldBeImmutable() {
        thrown.expect(UnsupportedOperationException.class);

        Board board = Board.builder().withPieces(singletonList(battleship)).build();

        board.getPieces().add(aircraftCarrier);
    }
}

File: game/src/test/java/com/bship/games/endpoints/cabinet/entity/BoardTest.kt
package com.bship.games.endpoints.cabinet.entity

import com.bship.games.endpoints.cabinet.entity.Piece.Companion.build
import com.bship.games.logic.definitions.Direction.DOWN
import com.bship.games.logic.definitions.Harbor
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test

class BoardTest {

    private lateinit var battleship: Piece
    private lateinit var aircraftCarrier: Piece

    @Before
    fun setup() {
        val start = Point(0, 0)
        val type = Harbor.BATTLESHIP
        battleship = build {
            withId { 1 }
            withType { type }
            withPlacement { start }
            withOrientation { DOWN }
        }

        val start1 = Point(1, 0)
        val type1 = Harbor.AIRCRAFT_CARRIER
        aircraftCarrier = build {
            withId { 1 }
            withType { type1 }
            withPlacement { start1 }
            withOrientation { DOWN }
        }
    }

    @Test
    fun addShip_shouldBeAbleToAddAShipToTheBoard() {
        val board = Board.build {
            withPieces { listOf(battleship) }
        }

        assertThat(board.pieces).contains(battleship)
    }

    @Test
    fun addShip_shouldBeAbleToAddMultipleShipsToTheBoard() {
        val board = Board.build { withPieces { listOf(battleship) } }
        val newBoard = board.copy { withPieces { board.pieces + aircraftCarrier } }

        assertThat(newBoard.pieces).contains(battleship, aircraftCarrier)
    }
}